### PR TITLE
Unit tests & corner case check for loading-getter

### PR DESCRIPTION
**Pull request by adamkaplan**
Added unit tests for existing code.
Tweaked docs to call out that the block is executing during full-lock – so I needs to be fast and can't call the cache again.

### CODE REVIEW

File: YMCache/YMMemoryCache.m [hunk 1]
```diff
@@ -191,6 +191,15 @@ - (id)objectForKey:(NSString *)key withDefault:(YMMemoryCacheObjectLoader)defaul
     if (!item) {
         __block id newItem;
         dispatch_barrier_sync(self.queue, ^{
+            // In order to ensure that read call are only blocking when they need
+            // to be, this mutating block is executed with it's own barrier. This
+            // means that we have a potential race condition if this method is called
+            // in parallel with the same missing key. Resolve by checking again!
+            newItem = self.items[key];
```
Comment (fsmanji) [inline YMCache/YMMemoryCache.m]: Great catch!
